yt_dlp/networking/impersonate.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass
+from typing import Any, Optional
+
+from .common import RequestHandler, register_preference
+from .exceptions import UnsupportedRequest
+from ..compat.types import NoneType
+from ..utils import classproperty
+from ..utils.networking import std_headers
+
+
+@dataclass(order=True)
```

Comment:
As I'm adding more browser versions in curl_cffi v0.6, using strings to represent browsers is becoming more and more cumbersome, do you mind if I incorporate this class into curl_cffi?

The benefits is that, for future versions, you may pass `impersonate=target` directly to curl_cffi, without mapping to strings like `chrome110`.

Reply:
The code is licensed under Unlicense, so feel free to take and mutate it.

Having a concrete structure to pass in will probably *not* simplify our code, since we would have to provide a simple way for extractors to use these. Since we don't require any packages to be installed, sth like `from curl_cffi import ImpersonationTarget` could not be used. Passing a string would be the simplest on the extractor developers. Just my two cents though, you should probably listen to @coletdjnz or @bashonly instead

Reply:
Agree w/Grub. My [rough implementation](https://github.com/coletdjnz/yt-dlp-dev/compare/networking/curl-impersonate...bashonly:yt-dlp:networking/curl-impersonate-ie) of the extractor impersonate API has the extractor helper methods accepting string args which are parsed the same as user input. Also afaik the `impersonate` module and `ImpersonationTarget` are intended to be able to work with request handlers other than curl_cffi in the future

Reply:
Yeah you can feel free to use it. As long as we can still generate a list of exact targets that can be passed to curl_cffi it will work for us.

Though, to give some background, it is designed specifically for our use case where:
- we want both a string and object representations of an impersonate target.
- we want to be able to specify a generic target (e.g just "chrome" instead of "chrome 110...") and let the handlers match (based on pre-defined preference)

Our framework is designed to abstract away the libraries like curl_cffi from the rest of the program, and have the ability to support multiple libraries (i.e not just curl_cffi). We'll still use it our end on any case.